### PR TITLE
Randomize apiserver timeout

### PR DESCRIPTION
**Pull request by bprashanth**
Connections get a random timeout between min/max instead of a hard 5 minute timout.
@lavalamp @fgrzadkowski @wojtek-t

### CODE REVIEW

**Comment by wojtek-t** [comment]
I think the proposal is great, I'm just wondering what the values of "minTimeout" and "maxTimeout" should be. Why did you choose 5 and 8 minutes - was it some guess or was it based on some data?

**Comment by bprashanth** [comment]
In practice I've seen the connection break at < 5m. This is something I'm currently tracking down.

That aside, I left the lower bound at whatever it is currently is. For the upper bound I observed a correlation between slow lists and > 30-50 concurrent lists, so I bucketed 100 nodes giving them each their own minute (didn't experiment more than that). I think anywhere between 6-8 should be Ok.

**Comment by wojtek-t** [comment]
I asked because my feeling is that maxTimeout = 10minutes = 2 * minTimeout would be more intuitive (we will be spreading the load "uniformly").
Is there any argument not to use 10 minutes? Is it too long? Or maybe we can use 4 minutes & 8 minutes?

**Comment by bprashanth** [comment]
Yeah, not sure but I think this is easily changable and requires some experimentation, so I can set it to 10 or 4 depending on what other voters prefer. The upper limit is a balance between how paranoid we want to be about dropping watch events vs how likely it is to happen on some set of nodes.

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -410,11 +420,12 @@ func (s *APIServer) Run(_ []string) error {
 	}
 
 	if secureLocation != "" {
+		timeout := time.Duration(MinTimeoutSecs+rand.Intn(MaxTimeoutSecs-MinTimeoutSecs)) * time.Second
```
Comment (wojtek-t) [inline cmd/kube-apiserver/app/server.go]: don't we want the same timeout for all http servers within one Kubelet?
I think it definitely makes sense to differentiate it between Kubelets, but maybe within a Kubelet they should be the same?

**Comment by lavalamp** [comment]
This LGTM.

**Comment by lavalamp** [comment]
These numbers seem fine at distributing load: http://play.golang.org/p/3FSbYtLmLT

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Why?

**Comment by wojtek-t** [comment]
Sorry - my comment above doesn't make much sense.

However - I'm afraid that I stopped understanding this PR. Basically, what you're doing with it is you are setting timeout at the apiserver level to be a random between 5 and 8 minutes. But it's still the same for all Kubelets. In other words, I think that instead of a burst after 5 minutes, we will have a burst after some random time between 5 and 8 minutes.
Am I missing something?

[What we would like to do is to setup the timeout at the request level in each Kubelet separately].

**Comment by bprashanth** [comment]
Hold on, @wojtek-t I need an associated timeout in each watch server.

**Comment by bprashanth** [comment]
No joken this time >.<
PTAL (still e2e-ing)

**Comment by wojtek-t** [comment]
Thanks - will take a look later today after my meetings.

File: cmd/kube-apiserver/app/server.go [hunk 2]
```diff
@@ -46,6 +46,12 @@ import (
 	"github.com/spf13/pflag"
 )
 
+const (
+	// Maximum duration before timing out read/write requests
+	// Set to a value larger than the timeouts in each watch server.
+	ReadWriteTimeoutMins = time.Minute * 60
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Remove "Mins" from name, it's a time.Duration, not a count of minutes.

File: pkg/apiserver/watch.go [hunk 3]
```diff
@@ -32,19 +34,48 @@ import (
...
+	if w.timeout == 0 {
+		return neverExitWatch, func() bool { return false }
+	}
+	t := time.NewTimer(w.timeout)
+	return t.C, t.Stop
+}
+
 // serveWatch handles serving requests to the server
 func serveWatch(watcher watch.Interface, scope RequestScope, w http.ResponseWriter, req *restful.Request) {
+
```
Comment (lavalamp) [inline pkg/apiserver/watch.go]: nit: extra blank line

File: pkg/apiserver/watch.go [hunk 4]
```diff
@@ -32,19 +34,48 @@ import (
...
+	// Each watch gets a random timeout to avoid thundering herds. Rand is seeded once in the api installer.
+	timeout := time.Duration(MinTimeoutSecs+rand.Intn(MaxTimeoutSecs-MinTimeoutSecs)) * time.Second
+
 	watchServer := &WatchServer{watcher, scope.Codec, func(obj runtime.Object) {
 		if err := setSelfLink(obj, req, scope.Namer); err != nil {
 			glog.V(5).Infof("Failed to set self link for object %v: %v", reflect.TypeOf(obj), err)
 		}
-	}}
+	}, &WatchTimeoutMonitor{timeout}}
```
Comment (lavalamp) [inline pkg/apiserver/watch.go]: Recommend naming the fields for clarity.

File: cmd/kube-apiserver/apiserver.go [hunk 5]
```diff
@@ -32,6 +34,8 @@ import (
 
 func main() {
 	runtime.GOMAXPROCS(runtime.NumCPU())
+	rand.Seed(time.Now().UTC().UnixNano())
```
Comment (lavalamp) [inline cmd/kube-apiserver/apiserver.go]: IMO it's not super important to seed rand, but it doesn't hurt either I suppose.

File: pkg/apiserver/watch.go [hunk 6]
```diff
@@ -32,19 +34,48 @@ import (
...
+	// nothing will ever be sent down this channel
+	neverExitWatch <-chan time.Time = make(chan time.Time)
+)
 
 func isWebsocketRequest(req *http.Request) bool {
 	return connectionUpgradeRegex.MatchString(strings.ToLower(req.Header.Get("Connection"))) && strings.ToLower(req.Header.Get("Upgrade")) == "websocket"
 }
 
+// TimeoutMonitor abstracts watch timeout logic for testing
+type TimeoutMonitor interface {
```
Comment (lavalamp) [inline pkg/apiserver/watch.go]: Can this be private?

Comment (lavalamp) [inline pkg/apiserver/watch.go]: This is a factory, right? It doesn't monitor anything, right? ```timeoutFactory``` would be much clearer name.

File: pkg/apiserver/watch.go [hunk 7]
```diff
@@ -32,19 +34,48 @@ import (
...
+	TimeoutCh() (<-chan time.Time, func() bool)
+}
+
+// WatchTimeoutMonitor implements TimeoutMonitor
+type WatchTimeoutMonitor struct {
```
Comment (lavalamp) [inline pkg/apiserver/watch.go]: ```realTimeoutFactory```?

**Comment by bprashanth** [comment]
Nits addressed, stil trying to run density on a 100 node cluster (it seems to be stuck polling for fluentd pods in some pre-test phase even on head today, will debug tomorrow) though e2e passed.

**Comment by bprashanth** [comment]
ok to test

**Comment by wojtek-t** [comment]
LGTM

**Pull request merged**